pkg/scheduler/queue/tree_queue_algo_querier_worker_queue_priority.go

```diff
@@ -0,0 +1,187 @@
+// This means that a tree will not always have all 4 node types for the 4 possible query component assignments.
+//
+// This has two implications for the distribution of workers across queue nodes:
+//  1. The modulo operation may modulo the worker ID by 1, 2, 3, or 4 depending on the number of node types
+//     currently present in the node order, which can change which node a worker ID is prioritized for.
+//  2. The node order changes as queues are deleted and re-created, so the worker ID to node mapping will change
+//     as the essentially random enqueue order places query component nodes in different positions in the order.
+//
+// We consider this a desirable property, as it ensures that a number of querier-workers which is not evenly
+// divisible by the number of query component nodes will, through the randomized changes in nodeOrder over time,
+// be distributed more evenly across the nodes than if length and order of the nodeOrder were fixed.
+//
+// Minimizing Idle Querier-Worker Capacity:
+// We say the queue nodes are "prioritized" for a worker rather than "assigned" to a worker
+// because the same worker ID is mapped to *start* at a certain queue node, but will move on to other nodes
+// if it cannot dequeue a request from any of the child queue nodes of its first prioritized queue node.
+// This can occur when this queue algorithm is placed at the highest layer of the tree and
+// the tenant-querier-shuffle-shard queue algorithm is placed at the second, leaf layer of the tree.
+// Ex:
+//
+//  1. The QuerierWorkerQueuePriorityAlgo begins with a nodeOrder of:
+//     ["ingester", "store-gateway", "ingester-and-store-gateway", "unknown"].
+//
+//  2. A querier-worker with workerID 0 requests to dequeue and is mapped to start with the "ingester" queue node.
+//
+//  3. The tree traversal algorithm recurs down to select child queue nodes of the "ingester" node,
+//     where each child queue node is a non-empty tenant-specific queue of ingester-only queries.
+//     The tenant-querier-shuffle-shard queue algorithm checks each tenant node for if it is sharded to this querier.
+//
+//  4. (a) The first tenant queue node found which is sharded to this querier will be dequeued from, and we are done.
+//
+//  4. (b) Otherwise, if none of those tenants are sharded to this querier, the tree traversal algorithm will return
+//     back up to the parent level and ask the QuerierWorkerQueuePriorityAlgo to select its next node.
+//     We continue to step 5.
+//
+//  5. The QuerierWorkerQueuePriorityAlgo will select the next node in the nodeOrder, "store-gateway".
+//     We return to step 3 and continue through steps 3, 4b, and 5, until we reach step 4a and exit.
+//
+// This process of continuing to search for requests to dequeue helps prevent querier-worker capacity from sitting idle
+// when there are no requests to dequeue for the query component node that the querier-worker was originally mapped to.
```

Comment:
```suggestion
// QuerierWorkerQueuePriorityAlgo implements QueuingAlgorithm, mapping worker IDs (assigned incrementally) to a queue
// node to prioritize. Querier-workers' priority is determined by currentQuerierWorker % len(nodeOrder), resulting in
// fair assignment of querier-workers across query component subtrees. Thus, when one query component is experiencing
// high latency (e.g., store-gateways), about 25% of querier-workers continue prioritizing queries for unaffected
// components.
//
// There are four possible query components: "ingester", "store-gateway", "ingester-and-store-gateway", and "unknown".
// When all four possible query component queue nodes exist, one out of four worker IDs starts dequeuing from any given
// queue node. For example, a querier-worker with ID 4 and node order ["ingester", "store-gateway",
// "ingester-and-store-gateway", "unknown"] first attempts to dequeue a query from the "ingester" node. However, it is
// not required that the number of querier-workers be divisible by four; this algorithm assumes that
// MultiQueuingAlgorithmTreeQueue always deletes empty leaf nodes and nodes with no children after a dequeue operation,
// and if a new request is enqueued for a previously-deleted paths, nodes are re-created. Thus:
//  1. The modulo operation may modulo the worker ID by 1, 2, 3, or 4 depending on the number of node types
//     currently present in the node order, which can change which node a worker ID is prioritized for.
//  2. The node order changes as queues are deleted and re-created, so the worker ID-to-node mapping changes as the
//     enqueue order places query component nodes in different positions in the order.
//
// These assumptions guarantee that even when the number of querier-workers is not evenly divisible by the number of
// query component nodes, through the randomized changes in node order over time, they are more likely to be evenly
// distributed across query component nodes than if length and order of those nodes were fixed.
//
// For a given worker, queue nodes are prioritized -- not assigned. Ceteris paribus, the same worker ID starts at a
// consistent queue node, but moves on to other nodes if it cannot dequeue a request from the subtree of its first
// prioritized queue node. This maximally utilizes querier-worker capacity when there are no process-able requests for
// a specific querier-worker and query component.
//
// A querier-worker can process queries for nodes it has not prioritized when this queue algorithm is applied at the
// highest layer of the tree and the tenant-querier-shuffle-shard queue algorithm applied at the second layer of the
// tree. If shuffle-sharding is enabled, a querier-worker that prioritizes ingester-only queries may not find
// ingester-only queries for any tenant it is assigned to, and move on to the next query component subtree. E.g.:
//
//  1. The QuerierWorkerQueuePriorityAlgo has node order:
//     ["ingester", "store-gateway", "ingester-and-store-gateway", "unknown"].
//
//  2. A querier-worker with workerID 0 requests to dequeue; it prioritizes the "ingester" queue node.
//
//  3. The dequeue operation recursively dequeues from "ingester" node. Each child node is a tenant-specific
//     queue of ingester-only queries. The tenantQuerierAssignments QueuingAlgorithm checks if any of its child nodes
//     (tenant queues) is assigned to this querier, and finds none.
//
//  4. We walk back up to the QuerierWorkerQueuePriorityAlgo level, not having dequeued anything. The
//     QuerierWorkerQueuePriorityAlgo increments currentNodeOrderIndex and selects the next node in nodeOrder (in
//     this example, "store-gateway"), and checks for dequeue-able queries again, from step 3, etc. until a
//     dequeue-able child is found, or every query component node has been checked for dequeue-able queries.
//
// TestMultiDimensionalQueueAlgorithmSlowConsumerEffects illustrates that this algorithm significantly outperforms
// a simple round-robin algorithm. Because round-robining selects a given query-component node ~25% of the time, when
// one query component is experiencing high latency, eventually, almost every querier-worker will be stuck processing
// queries for that query component.
```

Reply:
this was helpful - did a merge on our different wording and ordering here, and managed to cut the same number of lines (20) as your suggestions here.